Improve performance

apelisse opened the pull request:
Various improvements to allocations (up to 50% for update, 40% for FromUnstructured):
```
benchmark                                              old ns/op     new ns/op     delta
BenchmarkFieldSet/insert-20-12                         20178         19847         -1.64%
BenchmarkFieldSet/has-20-12                            721           710           -1.53%
BenchmarkFieldSet/serialize-20-12                      14198         14171         -0.19%
BenchmarkFieldSet/deserialize-20-12                    34374         35060         +2.00%
BenchmarkFieldSet/union-20-12                          11807         12268         +3.90%
BenchmarkFieldSet/intersection-20-12                   5041          5334          +5.81%
BenchmarkFieldSet/difference-20-12                     8918          9034          +1.30%
BenchmarkFieldSet/insert-50-12                         60155         59311         -1.40%
BenchmarkFieldSet/has-50-12                            829           813           -1.93%
BenchmarkFieldSet/serialize-50-12                      34321         33114         -3.52%
BenchmarkFieldSet/deserialize-50-12                    80336         79841         -0.62%
BenchmarkFieldSet/union-50-12                          30503         32520         +6.61%
BenchmarkFieldSet/intersection-50-12                   10552         10948         +3.75%
BenchmarkFieldSet/difference-50-12                     23098         23697         +2.59%
BenchmarkFieldSet/insert-100-12                        142519        145677        +2.22%
BenchmarkFieldSet/has-100-12                           1032          1010          -2.13%
BenchmarkFieldSet/serialize-100-12                     93235         90891         -2.51%
BenchmarkFieldSet/deserialize-100-12                   196329        193819        -1.28%
BenchmarkFieldSet/union-100-12                         52364         54112         +3.34%
BenchmarkFieldSet/intersection-100-12                  17147         16596         -3.21%
BenchmarkFieldSet/difference-100-12                    41190         41098         -0.22%
BenchmarkFieldSet/insert-500-12                        671973        686997        +2.24%
BenchmarkFieldSet/has-500-12                           1048          1130          +7.82%
BenchmarkFieldSet/serialize-500-12                     479285        465530        -2.87%
BenchmarkFieldSet/deserialize-500-12                   837894        828525        -1.12%
BenchmarkFieldSet/union-500-12                         236866        266410        +12.47%
BenchmarkFieldSet/intersection-500-12                  118440        116428        -1.70%
BenchmarkFieldSet/difference-500-12                    239103        215492        -9.87%
BenchmarkFieldSet/insert-1000-12                       1361049       1434539       +5.40%
BenchmarkFieldSet/has-1000-12                          1199          1164          -2.92%
BenchmarkFieldSet/serialize-1000-12                    989911        1037991       +4.86%
BenchmarkFieldSet/deserialize-1000-12                  1690249       1730177       +2.36%
BenchmarkFieldSet/union-1000-12                        450585        497112        +10.33%
BenchmarkFieldSet/intersection-1000-12                 191047        200137        +4.76%
BenchmarkFieldSet/difference-1000-12                   377929        407488        +7.82%
BenchmarkDeducedSimple-12                              117303        101274        -13.66%
BenchmarkDeducedNested-12                              318361        258506        -18.80%
BenchmarkDeducedNestedAcrossVersion-12                 317505        267320        -15.81%
BenchmarkLeafConflictAcrossVersion-12                  115084        98086         -14.77%
BenchmarkMultipleApplierRecursiveRealConversion-12     1649080       1585240       -3.87%
BenchmarkOperations/Pod/Update-12                      351283        176133        -49.86%
BenchmarkOperations/Pod/Apply-12                       516962        473135        -8.48%
BenchmarkOperations/Node/Update-12                     569627        238850        -58.07%
BenchmarkOperations/Node/Apply-12                      755372        697876        -7.61%
BenchmarkOperations/Endpoints/Update-12                56002         33171         -40.77%
BenchmarkOperations/Endpoints/Apply-12                 2081751       2082352       +0.03%
BenchmarkFromUnstructured/Pod-12                       202841        137831        -32.05%
BenchmarkFromUnstructured/Node-12                      327686        205776        -37.20%
BenchmarkFromUnstructured/Endpoints-12                 7909537       4781613       -39.55%

benchmark                                              old allocs     new allocs     delta
BenchmarkFieldSet/insert-20-12                         78             78             +0.00%
BenchmarkFieldSet/has-20-12                            2              2              +0.00%
BenchmarkFieldSet/serialize-20-12                      23             23             +0.00%
BenchmarkFieldSet/deserialize-20-12                    216            216            +0.00%
BenchmarkFieldSet/union-20-12                          53             53             +0.00%
BenchmarkFieldSet/intersection-20-12                   29             29             +0.00%
BenchmarkFieldSet/difference-20-12                     42             42             +0.00%
BenchmarkFieldSet/insert-50-12                         246            246            +0.00%
BenchmarkFieldSet/has-50-12                            2              2              +0.00%
BenchmarkFieldSet/serialize-50-12                      48             47             -2.08%
BenchmarkFieldSet/deserialize-50-12                    621            618            -0.48%
BenchmarkFieldSet/union-50-12                          134            135            +0.75%
BenchmarkFieldSet/intersection-50-12                   61             62             +1.64%
BenchmarkFieldSet/difference-50-12                     101            102            +0.99%
BenchmarkFieldSet/insert-100-12                        813            813            +0.00%
BenchmarkFieldSet/has-100-12                           3              3              +0.00%
BenchmarkFieldSet/serialize-100-12                     93             93             +0.00%
BenchmarkFieldSet/deserialize-100-12                   2158           2158           +0.00%
BenchmarkFieldSet/union-100-12                         258            257            -0.39%
BenchmarkFieldSet/intersection-100-12                  125            124            -0.80%
BenchmarkFieldSet/difference-100-12                    215            215            +0.00%
BenchmarkFieldSet/insert-500-12                        4140           4143           +0.07%
BenchmarkFieldSet/has-500-12                           3              3              +0.00%
BenchmarkFieldSet/serialize-500-12                     470            470            +0.00%
BenchmarkFieldSet/deserialize-500-12                   10787          10788          +0.01%
BenchmarkFieldSet/union-500-12                         1469           1469           +0.00%
BenchmarkFieldSet/intersection-500-12                  739            739            +0.00%
BenchmarkFieldSet/difference-500-12                    1162           1160           -0.17%
BenchmarkFieldSet/insert-1000-12                       8981           8977           -0.04%
BenchmarkFieldSet/has-1000-12                          3              3              +0.00%
BenchmarkFieldSet/serialize-1000-12                    998            998            +0.00%
BenchmarkFieldSet/deserialize-1000-12                  23496          23450          -0.20%
BenchmarkFieldSet/union-1000-12                        2847           2853           +0.21%
BenchmarkFieldSet/intersection-1000-12                 1325           1330           +0.38%
BenchmarkFieldSet/difference-1000-12                   2237           2238           +0.04%
BenchmarkDeducedSimple-12                              572            506            -11.54%
BenchmarkDeducedNested-12                              1562           1271           -18.63%
BenchmarkDeducedNestedAcrossVersion-12                 1630           1338           -17.91%
BenchmarkLeafConflictAcrossVersion-12                  544            493            -9.38%
BenchmarkMultipleApplierRecursiveRealConversion-12     5268           4903           -6.93%
BenchmarkOperations/Pod/Update-12                      843            404            -52.08%
BenchmarkOperations/Pod/Apply-12                       1179           1047           -11.20%
BenchmarkOperations/Node/Update-12                     1155           536            -53.59%
BenchmarkOperations/Node/Apply-12                      1876           1667           -11.14%
BenchmarkOperations/Endpoints/Update-12                100            73             -27.00%
BenchmarkOperations/Endpoints/Apply-12                 3143           3141           -0.06%
BenchmarkFromUnstructured/Pod-12                       477            321            -32.70%
BenchmarkFromUnstructured/Node-12                      1107           797            -28.00%
BenchmarkFromUnstructured/Endpoints-12                 21114          13075          -38.07%
```

apelisse commented:
Integrating this change in Kubernetes:
```
benchmark                                           old ns/op     new ns/op     delta
BenchmarkApplyNewObject//v1,_Kind=Pod-12            2292629       2271609       -0.92%
BenchmarkApplyNewObject//v1,_Kind=Node-12           4676879       4462379       -4.59%
BenchmarkApplyNewObject//v1,_Kind=Endpoints-12      89657544      88854655      -0.90%
BenchmarkUpdateNewObject//v1,_Kind=Pod-12           1115609       771370        -30.86%
BenchmarkUpdateNewObject//v1,_Kind=Node-12          1584057       1086912       -31.38%
BenchmarkUpdateNewObject//v1,_Kind=Endpoints-12     6047963       4073366       -32.65%
BenchmarkRepeatedUpdate-12                          959051        633715        -33.92%

benchmark                                           old allocs     new allocs     delta
BenchmarkApplyNewObject//v1,_Kind=Pod-12            8819           8325           -5.60%
BenchmarkApplyNewObject//v1,_Kind=Node-12           17696          16867          -4.68%
BenchmarkApplyNewObject//v1,_Kind=Endpoints-12      409590         401522         -1.97%
BenchmarkUpdateNewObject//v1,_Kind=Pod-12           3977           2982           -25.02%
BenchmarkUpdateNewObject//v1,_Kind=Node-12          5742           4290           -25.29%
BenchmarkUpdateNewObject//v1,_Kind=Endpoints-12     21441          13339          -37.79%
BenchmarkRepeatedUpdate-12                          3027           2026           -33.07%
```

apelisse commented:
cc @jennybuckley @lavalamp

lavalamp commented on value/value.go:
```diff
@@ -268,28 +265,17 @@ func (m *Map) Less(rhs *Map) bool {
-		f.Value = value
-		return
-	}
 	m.Items = append(m.Items, Field{Name: key, Value: value})
```
What prevents duplicates?

lavalamp commented on value/value.go:
```diff
@@ -268,28 +265,17 @@ func (m *Map) Less(rhs *Map) bool {
-		m.index = map[string]int{}
-		for i := range m.Items {
-			m.index[m.Items[i].Name] = i
+	for i := len(m.Items) - 1; i >= 0; i-- {
```
You may be optimizing for the one-shot case at the expense of the repeated case.

lavalamp commented on value/value.go:
```diff
@@ -334,25 +319,42 @@ func BooleanValue(b bool) Value {
 func (v Value) String() string {
```
This code should never be called from a performance-critical path.

apelisse replied:
We serialize pe's to string (to index maps), and pe's can be made of values.

apelisse commented:
I think I've addressed the comments, still the same imrpovements overall.

For Get, having the hashmap gets better for maps bigger than ~100 items, which are I think rare. We could do something better if needed for these case but I"m still hoping to find some other improvements.

apelisse commented:
Closing this in favor of something else ... TBD
